### DISCUSSION TITLE
How to close loop control my nema 23 12kg.cm motor using magnetic encoder ? any documents available ?

### DISCUSSION THREAD

**Discussion by hemangjoshi37a** [joshua-8/JMotor#122] [Q&A]
How to close loop control my nema 23 12kg.cm motor using magnetic encoder ? any documents available ? I am using esp32 as my controller.

**Comment by joshua-8** [comment]
Good morning @hemangjoshi37a,
Thanks for your interest in this library. 
This library has nothing for working with stepper motors (do you have more information about your motor? I'm assuming you have a stepper motor since they are commonly NEMA23). 
You could use this library to get measurements from [AS5048b](https://joshua-8.github.io/JMotor/class_j_encoder_a_s5048b_i2_c.html) and [AS5600](https://joshua-8.github.io/JMotor/class_j_encoder_a_s5600_i2_c.html) magnetic encoders but there's still a lot of code you would need to write to close the loop with a stepper motor.
Feel free to ask more questions, contribute code to this library, or show off whatever you end up building.
Good luck!

joshua-8

**Reply by hemangjoshi37a** [reply]
Definitely I myself am a big open source contributor and definitely if any improvements are made I will push.

**Reply by joshua-8** [reply]
thanks!

**Comment by hemangjoshi37a** [comment]
I have these components as list :
1 : ESP32 as controller 
2 : Nema 23 4-wire 12kg.cm stepper motor 
3 : DM542 Stepper motor driver
4 : AS5600 magnetic encoder

**Reply by joshua-8** [reply]
1. ESP32 microcontrollers are great, good choice
2. adding stepper motors to this library would be a lot of work, but if you can do it that would be great
3. looks like a nice motor driver
4. That's a good magnetic encoder. If you want an alternative library try this: https://github.com/RobTillaart/AS5600

**Comment by hemangjoshi37a** [comment]
Basically I want to have a function in which we can write `MoveTo(560)` where 560 is degree .

**Reply by joshua-8** [reply]
Do you need to use closed loop control with an encoder? Stepper motors can move precise distances without any sensors. I've had good luck with this library for controlling stepper motors, it has functions to go to positions: https://www.arduino.cc/reference/en/libraries/accelstepper/

**Reply by hemangjoshi37a** [reply]
actually stepper motors are great for precise control until they miss any step. for longer running/rotational operations for example for 5 rotation or 10 rotations it sometimes gives non repeatable movements with very large variations and I have consulted with many of my friends about this and most of them say that closed loop is the ultimate way for this . I run company names https://hjLabs.in and we design stepper and servo motor around industrial automation applications .

**Reply by joshua-8** [reply]
Cool! I'm looking forward to hearing how you add closed loop control to a stepper motor.
